File: Sources/Device.swift
import Foundation

let kBrainBitChannels = ["O1", "O2", "T3", "T4"]
let kBrainBitSampleRateHz = 250

/// Scan for BrainBit headbands for `seconds`.
/// The returned scanner is still alive — keep a reference to it for the
/// lifetime of any sensors you create from it.
func scanForBrainBits(seconds: TimeInterval = 5.0) -> (NTScanner, [NTSensorInfo]) {
    let scanner = NTScanner(sensorFamily: [
        NSNumber(value: NTSensorFamily.leBrainBit.rawValue)
    ])!

    var seen: [String: NTSensorInfo] = [:]
    let lock = NSLock()

    scanner.setSensorsCallback { (sensors: [NTSensorInfo]) in
        lock.lock(); defer { lock.unlock() }
        for s in sensors { seen[s.address] = s }
    }

    scanner.startScan()
    Thread.sleep(forTimeInterval: seconds)
    scanner.stopScan()
    scanner.setSensorsCallback(nil)

    if let cached = scanner.sensors {
        lock.lock()
        for s in cached { seen[s.address] = s }
        lock.unlock()
    }

    return (scanner, Array(seen.values))
}

/// Pick the SensorInfo matching `serial`, or the first one if `serial` is nil.
func pickHeadband(_ sensors: [NTSensorInfo], serial: String?) -> NTSensorInfo? {
    guard !sensors.isEmpty else { return nil }
    guard let serial = serial else { return sensors[0] }
    return sensors.first { $0.serialNumber == serial }
}

/// Scan, pick a device, create+connect the sensor. Caller is responsible for
/// holding the returned scanner alive and calling Disconnect when done.
func connectBrainBit(serial: String?, scanSeconds: TimeInterval = 5.0) throws -> (NTScanner, NTBrainBit, NTSensorInfo) {
    let (scanner, sensors) = scanForBrainBits(seconds: scanSeconds)
    guard let info = pickHeadband(sensors, serial: serial) else {
        throw EEGError.noDevice(serial: serial, scanSeconds: scanSeconds)
    }
    let sensor = scanner.createSensor(info)
    guard let bb = sensor as? NTBrainBit else {
        throw EEGError.wrongDeviceType(name: info.name)
    }
    return (scanner, bb, info)
}

enum EEGError: Error, CustomStringConvertible {
    case noDevice(serial: String?, scanSeconds: TimeInterval)
    case wrongDeviceType(name: String)

    var description: String {
        switch self {
        case .noDevice(let serial, let s):
            let suffix = serial.map { " with serial \($0)" } ?? ""
            return "No BrainBit found\(suffix) after \(Int(s))s scan."
        case .wrongDeviceType(let name):
            return "Connected to \(name) but it didn't expose the BrainBit interface."
        }
    }
}

File: src/_device.py
"""Shared helpers: scan for BrainBit headbands and connect to one."""
from __future__ import annotations

import time
from typing import Optional

from neurosdk.scanner import Scanner
from neurosdk.cmn_types import SensorFamily


SCAN_SECONDS_DEFAULT = 5.0


def scan(seconds: float = SCAN_SECONDS_DEFAULT) -> tuple[Scanner, list]:
    """Run a BLE scan and return (scanner, sensors_seen).

    The scanner is returned still alive so the caller can call
    ``scanner.create_sensor(...)`` on one of the SensorInfo entries.
    """
    scanner = Scanner([SensorFamily.LEBrainBit])
    seen: dict[str, object] = {}

    def on_change(_scanner, sensors):
        for s in sensors:
            seen[s.Address] = s

    scanner.sensorsChanged = on_change
    scanner.start()
    time.sleep(seconds)
    scanner.stop()
    scanner.sensorsChanged = None

    # Merge anything the SDK already had cached.
    for s in scanner.sensors():
        seen[s.Address] = s

    return scanner, list(seen.values())


def pick(sensors, serial: Optional[str] = None):
    """Pick the SensorInfo matching ``serial``, or the first one if not given."""
    if not sensors:
        return None
    if serial is None:
        return sensors[0]
    for s in sensors:
        if s.SerialNumber == serial:
            return s
    return None


def connect(serial: Optional[str] = None, scan_seconds: float = SCAN_SECONDS_DEFAULT):
    """Scan, pick a device, return (scanner, sensor, sensor_info).

    Raises RuntimeError if no matching device is found. The scanner must be
    kept alive for the lifetime of the sensor — return it so callers can hold
    a reference and clean up at the end.
    """
    scanner, sensors = scan(scan_seconds)
    info = pick(sensors, serial)
    if info is None:
        scanner = None  # let it dispose
        raise RuntimeError(
            f"No BrainBit found"
            + (f" with serial {serial}" if serial else "")
            + f" after {scan_seconds:.0f}s scan."
        )
    sensor = scanner.create_sensor(info)
    return scanner, sensor, info
